satellite/satellitedb/usercredits.go

```diff
@@ -0,0 +1,172 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"context"
+	"strconv"
+	"time"
+
+	"github.com/skyrings/skyring-common/tools/uuid"
+	"github.com/zeebo/errs"
+
+	"storj.io/storj/satellite/console"
+	dbx "storj.io/storj/satellite/satellitedb/dbx"
+)
+
+type usercredits struct {
+	db *dbx.DB
+}
+
+type updateInfo struct {
+	id      int
+	credits int
+}
+
+// TotalReferredCount returns the total amount of referral a user has made based on user id
+func (c *usercredits) TotalReferredCount(ctx context.Context, id uuid.UUID) (int64, error) {
+	totalReferred, err := c.db.Count_UserCredit_By_ReferredBy(ctx, dbx.UserCredit_ReferredBy(id[:]))
+	if err != nil {
+		return totalReferred, errs.Wrap(err)
+	}
+
+	return totalReferred, nil
+}
+
+// GetAvailableCredits returns all records of user credit that are not expired or used
+func (c *usercredits) GetAvailableCredits(ctx context.Context, referrerID uuid.UUID, expirationEndDate time.Time) ([]console.UserCredit, error) {
+	availableCredits, err := c.db.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(referrerID[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return fromDBX(availableCredits)
+}
+
+// Create insert a new record of user credit
+func (c *usercredits) Create(ctx context.Context, userCredit console.UserCredit) (*console.UserCredit, error) {
+	createdUser, err := c.db.Create_UserCredit(ctx,
+		dbx.UserCredit_UserId(userCredit.UserID[:]),
+		dbx.UserCredit_OfferId(userCredit.OfferID),
+		dbx.UserCredit_CreditsEarnedInCents(userCredit.CreditsEarnedInCents),
+		dbx.UserCredit_ExpiresAt(userCredit.ExpiresAt),
+		dbx.UserCredit_Create_Fields{
+			ReferredBy: dbx.UserCredit_ReferredBy(userCredit.ReferredBy[:]),
+		},
+	)
+	if err != nil {
+		return nil, errs.Wrap(err)
+	}
+
+	return convertDBX(createdUser)
+}
+
+// UpdateAvailableCredits updates user's available credits based on their spending and the time of their spending
+func (c *usercredits) UpdateAvailableCredits(ctx context.Context, creditsToCharge int, id uuid.UUID, expirationEndDate time.Time) (remainingCharge int, err error) {
+	tx, err := c.db.Open(ctx)
+	if err != nil {
+		return creditsToCharge, errs.Wrap(err)
+	}
+
+	availableCredits, err := tx.All_UserCredit_By_UserId_And_ExpiresAt_Greater_And_CreditsUsedInCents_Less_CreditsEarnedInCents_OrderBy_Asc_ExpiresAt(ctx,
+		dbx.UserCredit_UserId(id[:]),
+		dbx.UserCredit_ExpiresAt(expirationEndDate),
+	)
+	if err != nil {
+		return creditsToCharge, errs.Wrap(errs.Combine(err, tx.Rollback()))
+	}
+	if len(availableCredits) == 0 {
+		return creditsToCharge, errs.Combine(errs.New("No available credits"), tx.Commit())
+	}
+
+	var infos []updateInfo
+	remainingCharge = creditsToCharge
+	for _, credit := range availableCredits {
```

Comment:
What happens if there are many concurrent queries? Can this be written without transactions?

Reply:
The update happens after a user has paid their bill. Will there be a scenario that user pays multiple times in a row?

Reply:
Unlikely. But, I would like as much of the code work with concurrent requests... it helps to avoid debugging weird scenarios.

Reply:
It seems like this could be done as a single query for Postgres, leaving tx for sqlite won't be as much of a problem.

Reply:
but i guess we can leave it as a tx for now.

Reply:
I had to make separate queries for `sqlite3` and `postgres` to use `statement` to pass in values. Maybe I will try to just use single query for Postgres and leave the sqlite for tx